README.md
<p align="center">
    <a href="https://github.com/yii1tech" target="_blank">
        <img src="https://avatars.githubusercontent.com/u/134691944" height="100px">
    </a>
    <h1 align="center">Model Attributes Typecast Extension for Yii 1</h1>
    <br>
</p>

This extension provides support for Yii1 Model and ActiveRecord attributes typecast.

For license information check the [LICENSE](LICENSE.md)-file.

[![Latest Stable Version](https://img.shields.io/packagist/v/yii1tech/model-typecast.svg)](https://packagist.org/packages/yii1tech/model-typecast)
[![Total Downloads](https://img.shields.io/packagist/dt/yii1tech/model-typecast.svg)](https://packagist.org/packages/yii1tech/model-typecast)
[![Build Status](https://github.com/yii1tech/model-typecast/workflows/build/badge.svg)](https://github.com/yii1tech/model-typecast/actions)


Installation
------------

The preferred way to install this extension is through [composer](http://getcomposer.org/download/).

Either run

```
php composer.phar require --prefer-dist yii1tech/model-typecast
```

or add

```json
"yii1tech/model-typecast": "*"
```

to the "require" section of your composer.json.


Usage
-----

This extension provides support for Yii1 Model and ActiveRecord automatic attributes typecast.
It is performed via usage of `\yii1tech\model\typecast\TypecastBehavior` behavior.
It should be attached to `\CModel` or `\CActiveRecord` descendant.
For example:

```php
<?php

use yii1tech\model\typecast\TypecastBehavior;

class Item extends CActiveRecord
{
    public function behaviors()
    {
        return [
            'typecastBehavior' => [
                'class' => TypecastBehavior::class,
                'attributeTypes' => [
                    'id' => TypecastBehavior::TYPE_INTEGER,
                    'amount' => TypecastBehavior::TYPE_INTEGER,
                    'price' => TypecastBehavior::TYPE_FLOAT,
                    'is_active' => TypecastBehavior::TYPE_BOOLEAN,
                    'created_at' => TypecastBehavior::TYPE_DATETIME,
                    'json_data' => TypecastBehavior::TYPE_ARRAY_OBJECT,
                ],
                'typecastAfterValidate' => true,
                'typecastBeforeSave' => false,
                'typecastAfterSave' => true,
                'typecastAfterFind' => true,
            ],
        ];
    }

    // ...
}
```

> Tip: you may leave `\yii1tech\model\typecast\TypecastBehavior::$attributeTypes` blank - in this case its value will be detected
automatically: for ActiveRecord - based on owner DB table schema, for regular model - based validation rules.

In the above example attribute typecasting will be automatically performed in following cases:

- after model successful validation
- after model successful saving
- after model retrieval from Database

For example:

```php
<?php

$model = new Item();
$model->setAttributes([
    'name' => 'item name',
    'price' => '10.50',
    'amount' => '14',
    'is_active' => '1',
]);

if ($model->validate()) {
    var_dump($model->id); // outputs: int(123456)
    var_dump($model->price); // outputs: float(10.5)
    var_dump($model->amount); // outputs: int(14)
    var_dump($model->is_active); // outputs: bool(true)
}

$model = Item::model()->findByPk($id);
var_dump($model->id); // outputs: int(12345)
var_dump($model->amount); // outputs: int(18)
var_dump($model->is_active); // outputs: bool(true)
```

You can manually trigger attribute typecasting anytime invoking `\yii1tech\model\typecast\TypecastBehavior::typecastAttributes()` method:

```php
<?php

$model = new Item();
$model->price = '38.5';
$model->is_active = 1;
$model->typecastAttributes();

var_dump($model->price); // outputs: float(38.5)
var_dump($model->is_active); // outputs: bool(true)
```


### JSON Typecasting <span id="json-typecasting"></span>

This behavior allows automatic conversion of array or traversable objects into JSON string on model saving.
For example:

```php
<?php

$model = new Item();
$model->json_data = [ // will be saved in DB as '{foo: "bar"}'
    'foo' => 'bar',
];
$model->save();
```

> Note: such conversion will take place even, if there is no direct attribute type specification. 

You can typecast JSON column value either to plain `array` or `\ArrayObject` instance.
Plain arrays consume less memory, but writing of its particular internal keys will not work.
`\ArrayObject` allows free operation over internal keys, but you should note that its value always passed by reference.

```php
<?php

// in case mapping in `attributeTypes` is set to `TypecastBehavior::TYPE_ARRAY`
use yii1tech\model\typecast\TypecastBehavior;

class Item extends CActiveRecord
{
    public function behaviors()
    {
        return [
            'typecastBehavior' => [
                'class' => TypecastBehavior::class,
                'attributeTypes' => [
                    'json_data' => TypecastBehavior::TYPE_ARRAY,
                ],
            ],
        ];
    }

    // ...
}

$model = Item::model()->findByPk($id);
var_dump($model->json_data); // outputs: array(1) {...}
var_dump($model->json_data['foo']); // outputs: string(bar)
$model->json_data['foo'] = 'new value'; // PHP E_NOTICE: Indirect modification of overloaded property Item::$json_data has no effect!
$model->json_data = [ // no problem
    'foo' => 'new value',
];
$model->save();

// in case mapping in `attributeTypes` is set to `TypecastBehavior::TYPE_ARRAY_OBJECT`
use yii1tech\model\typecast\TypecastBehavior;

class Item extends CActiveRecord
{
    public function behaviors()
    {
        return [
            'typecastBehavior' => [
                'class' => TypecastBehavior::class,
                'attributeTypes' => [
                    'json_data' => TypecastBehavior::TYPE_ARRAY_OBJECT,
                ],
            ],
        ];
    }

    // ...
}

$model = Item::model()->findByPk($id);
var_dump($model->json_data); // outputs: object(ArrayObject) {...}
var_dump($model->json_data['foo']); // outputs: string(bar)
$model->json_data['foo'] = 'new value'; // no problem
$jsonDataCopy = $model->json_data; // new variable holds the reference to `\ArrayObject` instance!
$jsonDataCopy['foo'] = 'value from copy'; // changes value of `$model->json_data`!
var_dump($model->json_data['foo']); // outputs: string(value from copy)
$model->save();
```


### DateTime Typecasting <span id="datetime-typecasting"></span>

This behavior allows automatic conversion of `\DateTime` instances into ISO datetime string on model saving.
For example:

```php
<?php

use yii1tech\model\typecast\TypecastBehavior;

class Item extends CActiveRecord
{
    public function behaviors()
    {
        return [
            'typecastBehavior' => [
                'class' => TypecastBehavior::class,
                'attributeTypes' => [
                    'created_at' => TypecastBehavior::TYPE_DATETIME,
                ],
            ],
        ];
    }

    // ...
}

$model = new Item();
$model->created_at = new DateTime('now'); // will be saved in DB as '2023-12-22 10:14:17'
$model->save();

$model = Item::model()->findByPk($id);
var_dump($model->created_at); // outputs: object(DateTime)
```

In case you store the dates using integer Unix timestamp, you can use `\yii1tech\model\typecast\TypecastBehavior::TYPE_TIMESTAMP` for correct
conversion. For example:

```php
<?php

use yii1tech\model\typecast\TypecastBehavior;

class Item extends CActiveRecord
{
    public function behaviors()
    {
        return [
            'typecastBehavior' => [
                'class' => TypecastBehavior::class,
                'attributeTypes' => [
                    'created_at' => TypecastBehavior::TYPE_TIMESTAMP,
                ],
            ],
        ];
    }

    // ...
}

$model = new Item();
$model->created_at = new DateTime('now'); // will be saved in DB as '1703257478'
$model->save();

$model = Item::model()->findByPk($id);
var_dump($model->created_at); // outputs: object(DateTime)
```

This extension also supports [nesbot/carbon](https://packagist.org/packages/nesbot/carbon) package.
In order to convert dates to `\Carbon\Carbon` you should use following types:

- `\yii1tech\model\typecast\TypecastBehavior::TYPE_DATETIME_CARBON`
- `\yii1tech\model\typecast\TypecastBehavior::TYPE_TIMESTAMP_CARBON`


### Custom Typecasting <span id="custom-typecasting"></span>

You may specify any custom typecasting for the attribute using a callable as a type specification at `\yii1tech\model\typecast\TypecastBehavior::$attributeTypes`.
For example:

```php
<?php

use yii1tech\model\typecast\TypecastBehavior;

class Item extends CActiveRecord
{
    public function behaviors()
    {
        return [
            'typecastBehavior' => [
                'class' => TypecastBehavior::class,
                'attributeTypes' => [
                    'heap_data' => function ($value) {
                        if (is_object($value)) {
                            return $value;
                        }
                        
                        $heap = new \SplMaxHeap();
                        foreach (json_decode($value) as $element) {
                            $heap->insert($element);
                        }
                        
                        return $heap;
                    },
                ],
            ],
        ];
    }

    // ...
}

$model = Item::model()->findByPk($id);
var_dump($model->heap_data); // outputs: object(SplMaxHeap)
```


CHANGELOG.md
Yii1 Model Attribute Typecast extension
=======================================

1.0.0, December 22, 2023
------------------------

- Initial release.


tests/TestCase.php
<?php

namespace yii1tech\model\typecast\test;

use CConsoleApplication;
use CMap;
use Yii;
use yii1tech\model\typecast\TypecastBehavior;

class TestCase extends \PHPUnit\Framework\TestCase
{
    /**
     * {@inheritdoc}
     */
    protected function setUp(): void
    {
        parent::setUp();

        $this->mockApplication();

        $this->setupTestDbData();
    }

    /**
     * {@inheritdoc}
     */
    protected function tearDown(): void
    {
        TypecastBehavior::clearAutoDetectedAttributeTypes();

        $this->destroyApplication();

        parent::tearDown();
    }

    /**
     * Populates Yii::app() with a new application
     * The application will be destroyed on tearDown() automatically.
     * @param array $config The application configuration, if needed
     * @param string $appClass name of the application class to create
     */
    protected function mockApplication($config = [], $appClass = CConsoleApplication::class)
    {
        Yii::setApplication(null);

        new $appClass(CMap::mergeArray([
            'id' => 'testapp',
            'basePath' => __DIR__,
            'components' => [
                'db' => [
                    'class' => \CDbConnection::class,
                    'connectionString' => 'sqlite::memory:',
                ],
                'cache' => [
                    'class' => \CDummyCache::class,
                ],
            ],
        ], $config));
    }

    /**
     * Destroys Yii application by setting it to null.
     */
    protected function destroyApplication()
    {
        Yii::setApplication(null);
    }

    /**
     * Setup tables for test ActiveRecord
     */
    protected function setupTestDbData()
    {
        $db = Yii::app()->getDb();

        // Structure :
        $db->createCommand()
            ->createTable('item', [
                'id' => 'pk',
                'category_id' => 'integer',
                'name' => 'string',
                'price' => 'float',
                'is_active' => 'boolean DEFAULT 0',
                'created_timestamp' => 'integer',
                'created_date' => 'datetime',
                'callback' => 'string',
                'data_array' => 'json',
                'data_array_object' => 'json',
            ]);

        // Data :
        $builder = $db->getCommandBuilder();

        $builder->createMultipleInsertCommand('item', [
            [
                'category_id' => 1,
                'name' => 'item1',
                'is_active' => 0,
                'created_timestamp' => time(),
                'created_date' => date('Y-m-d H:i:s'),
            ],
            [
                'category_id' => 2,
                'name' => 'item2',
                'is_active' => 1,
                'created_timestamp' => time(),
                'created_date' => date('Y-m-d H:i:s'),
            ],
        ])->execute();
    }
}

tests/data/Item.php
<?php

namespace yii1tech\model\typecast\test\data;

use CActiveRecord;

/**
 * @property int $id
 * @property int $category_id
 * @property string $name
 * @property float $price
 * @property bool $is_active
 * @property int $created_timestamp
 * @property string $created_date
 * @property string $callback
 * @property array|string $data_array
 * @property \ArrayObject|string $data_array_object
 */
class Item extends CActiveRecord
{
    /**
     * {@inheritdoc}
     */
    public static function model($className = __CLASS__)
    {
        return parent::model($className);
    }

    /**
     * {@inheritdoc}
     */
    public function tableName(): string
    {
        return 'item';
    }
}

src/TypecastBehavior.php
<?php

namespace yii1tech\model\typecast;

use CActiveRecord;
use Carbon\Carbon;
use CBehavior;
use CBooleanValidator;
use CDbColumnSchema;
use CDbException;
use CEvent;
use CModelEvent;
use CNumberValidator;
use CStringValidator;
use InvalidArgumentException;
use Yii;

/**
 * TypecastBehavior provides an ability of automatic model attribute typecasting.
 *
 * This behavior should be attached to {@see \CModel} or {@see \CActiveRecord} descendant.
 *
 * For example:
 *
 * ```php
 * use yii1tech\model\typecast\TypecastBehavior;
 *
 * class Item extends CActiveRecord
 * {
 *     public function behaviors()
 *     {
 *         return [
 *             'typecastBehavior' => [
 *                 'class' => TypecastBehavior::class,
 *                 'attributeTypes' => [
 *                     'id' => TypecastBehavior::TYPE_INTEGER,
 *                     'amount' => TypecastBehavior::TYPE_INTEGER,
 *                     'price' => TypecastBehavior::TYPE_FLOAT,
 *                     'is_active' => TypecastBehavior::TYPE_BOOLEAN,
 *                     'created_at' => TypecastBehavior::TYPE_DATETIME,
 *                     'json_data' => TypecastBehavior::TYPE_ARRAY_OBJECT,
 *                 ],
 *                 'typecastAfterValidate' => true,
 *                 'typecastBeforeSave' => false,
 *                 'typecastAfterSave' => true,
 *                 'typecastAfterFind' => true,
 *             ],
 *         ];
 *     }
 *
 *     // ...
 * }
 * ```
 *
 * Tip: you may leave {@see $attributeTypes} blank - in this case its value will be detected
 * automatically: for ActiveRecord - based on owner DB table schema, for regular model - based validation rules.
 *
 * Note: you can manually trigger attribute typecasting anytime invoking {@see typecastAttributes()} method:
 *
 * ```php
 * $model = new Item();
 * $model->price = '38.5';
 * $model->is_active = 1;
 * $model->typecastAttributes();
 * ```
 *
 * This behavior allows automatic conversion of {@see \DateTime} instances into ISO datetime string, and array into JSON string
 * on model saving. For example:
 *
 * ```php
 * $model = new Item();
 * $model->created_at = new DateTime('now'); // will be saved in DB as '2023-12-22 10:14:17'
 * $model->json_data = [ // will be saved in DB as '{foo: "bar"}'
 *     'foo' => 'bar',
 * ];
 * $model->save();
 * ```
 *
 * @property \CModel|\CActiveRecord $owner The owner component that this behavior is attached to.
 *
 * @author Paul Klimov <klimov.paul@gmail.com>
 * @since 1.0
 */
class TypecastBehavior extends CBehavior
{
    /**
     * Converts attribute to `int`.
     */
    const TYPE_INTEGER = 'integer';
    /**
     * Converts attribute to `float`.
     */
    const TYPE_FLOAT = 'float';
    /**
     * Converts attribute to `bool`.
     */
    const TYPE_BOOLEAN = 'boolean';
    /**
     * Converts attribute to `string`.
     */
    const TYPE_STRING = 'string';
    /**
     * Converts JSON to array and vice versa.
     */
    const TYPE_ARRAY = 'array';
    /**
     * Converts JSON to {@see \ArrayObject} and vice versa.
     */
    const TYPE_ARRAY_OBJECT = 'array-object';
    /**
     * Converts ISO datetime string into {@see \DateTime} and vice versa.
     */
    const TYPE_DATETIME = 'datetime';
    /**
     * Converts integer Unix timestamp into {@see \DateTime} and vice versa.
     */
    const TYPE_TIMESTAMP = 'timestamp';
    /**
     * Converts ISO datetime string into {@see \Carbon\Carbon} and vice versa.
     * @see https://packagist.org/packages/nesbot/carbon
     */
    const TYPE_DATETIME_CARBON = 'datetime-carbon';
    /**
     * Converts integer Unix timestamp into {@see \Carbon\Carbon} and vice versa.
     * @see https://packagist.org/packages/nesbot/carbon
     */
    const TYPE_TIMESTAMP_CARBON = 'timestamp-carbon';

    /**
     * @var array<string, string|callable>|null attribute typecast map in format: attributeName => type.
     * Type can be set via PHP callable, which accept raw value as an argument and should return
     * typecast result.
     * For example:
     *
     * ```php
     * [
     *     'amount' => 'integer',
     *     'price' => 'float',
     *     'is_active' => 'boolean',
     *     'date' => function ($value) {
     *         return ($value instanceof \DateTime) ? $value->getTimestamp(): (int) $value;
     *     },
     * ]
     * ```
     *
     * If not set, attribute type map will be composed automatically from the owner validation rules.
     */
    public $attributeTypes;
    /**
     * @var bool whether to skip typecasting of `null` values.
     * If enabled attribute value which equals to `null` will not be type-casted (e.g. `null` remains `null`),
     * otherwise it will be converted according to the type configured at {@see attributeTypes}.
     */
    public $skipOnNull = true;
    /**
     * @var bool whether to perform typecasting after owner model validation.
     * Note that typecasting will be performed only if validation was successful, e.g. owner model has no errors.
     * Note that changing this option value will have no effect after this behavior has been attached to the model.
     */
    public $typecastAfterValidate = true;
    /**
     * @var bool whether to perform typecasting before saving owner model (insert or update).
     * This option may be disabled in order to achieve better performance.
     * For example, in case of {@see \CActiveRecord} usage, typecasting before save
     * will grant no benefit and thus can be disabled.
     */
    public $typecastBeforeSave = false;
    /**
     * @var bool whether to perform typecasting after saving owner model (insert or update).
     * This option may be disabled in order to achieve better performance.
     * For example, in case of {@see \CActiveRecord} usage, typecasting after save
     * will grant no benefit and thus can be disabled.
     * @since 2.0.14
     */
    public $typecastAfterSave = true;
    /**
     * @var bool whether to perform typecasting after retrieving owner model data from
     * the database (after find or refresh).
     * This option may be disabled in order to achieve better performance.
     * For example, in case of {@see \CActiveRecord} usage, typecasting after find
     * will grant no benefit in most cases and thus can be disabled.
     * Note that changing this option value will have no effect after this behavior has been attached to the model.
     */
    public $typecastAfterFind = true;

    /**
     * @var array<string, mixed> stashed raw attributes, used to transfer raw non-scalar values from {@see beforeSave()} to {@see afterSave()}.
     */
    private $_stashedAttributes = [];

    /**
     * @var array<string, array> internal static cache for auto detected {@see $attributeTypes} values
     * in format: `ownerClassName => attributeTypes`.
     */
    private static $autoDetectedAttributeTypes = [];

    /**
     * {@inheritdoc}
     */
    public function attach($owner): void
    {
        parent::attach($owner);

        if ($this->attributeTypes === null) {
            $this->attributeTypes = $this->detectAttributeTypes();
        }
    }

    /**
     * Detects (guesses) the attribute types analysing owner class.
     *
     * @return array<string, string> detected attribute types.
     */
    protected function detectAttributeTypes(): array
    {
        $ownerClass = get_class($this->owner);
        if (!isset(self::$autoDetectedAttributeTypes[$ownerClass])) {
            if ($this->owner instanceof CActiveRecord) {
                self::$autoDetectedAttributeTypes[$ownerClass] = $this->detectAttributeTypesFromSchema();
            } else {
                self::$autoDetectedAttributeTypes[$ownerClass] = $this->detectAttributeTypesFromRules();
            }
        }

        return self::$autoDetectedAttributeTypes[$ownerClass];
    }

    /**
     * Clears internal static cache of auto-detected {@see $attributeTypes} values
     * over all affected owner classes.
     */
    public static function clearAutoDetectedAttributeTypes(): void
    {
        self::$autoDetectedAttributeTypes = [];
    }

    /**
     * Typecast owner attributes according to {@see $attributeTypes}.
     *
     * @param array|null $attributeNames list of attribute names that should be type-casted.
     * If this parameter is empty, it means any attribute listed in the {@see $attributeTypes}
     * should be type-casted.
     * @return \CModel|\CActiveRecord owner instance.
     */
    public function typecastAttributes($attributeNames = null)
    {
        $attributeTypes = [];

        if ($attributeNames === null) {
            $attributeTypes = $this->attributeTypes;
        } else {
            foreach ($attributeNames as $attribute) {
                if (!isset($this->attributeTypes[$attribute])) {
                    throw new InvalidArgumentException("There is no type mapping for '{$attribute}'.");
                }
                $attributeTypes[$attribute] = $this->attributeTypes[$attribute];
            }
        }

        foreach ($attributeTypes as $attribute => $type) {
            $value = $this->owner->{$attribute};
            if ($this->skipOnNull && $value === null) {
                continue;
            }
            $this->owner->{$attribute} = $this->typecastValue($value, $type);
        }

        return $this->owner;
    }

    /**
     * Casts the given value to the specified type.
     *
     * @param mixed $value value to be type-casted.
     * @param string|callable $type type name or typecast callable.
     * @return mixed typecast result.
     */
    protected function typecastValue($value, $type)
    {
        if (!is_scalar($type)) {
            return call_user_func($type, $value);
        }

        switch ($type) {
            case self::TYPE_INTEGER:
            case 'int':
                return (int) $value;
            case self::TYPE_FLOAT:
                return (float) $value;
            case self::TYPE_BOOLEAN:
            case 'bool':
                return (bool) $value;
            case self::TYPE_STRING:
                return (string) $value;
            case self::TYPE_ARRAY:
                if (empty($value) || is_iterable($value)) {
                    return $value;
                }

                return json_decode($value, true);
            case self::TYPE_ARRAY_OBJECT:
                if (empty($value) || is_iterable($value)) {
                    return $value;
                }

                return new \ArrayObject(json_decode($value, true));
            case self::TYPE_DATETIME:
                if ($value === null || $value instanceof \DateTime) {
                    return $value;
                }

                return \DateTime::createFromFormat('Y-m-d H:i:s', (string) $value);
            case self::TYPE_TIMESTAMP:
                if ($value === null || $value instanceof \DateTime) {
                    return $value;
                }

                return (new \DateTime())->setTimestamp((int) $value);
            case self::TYPE_DATETIME_CARBON:
                if ($value === null || $value instanceof \DateTime) {
                    return $value;
                }

                if (!class_exists(Carbon::class)) {
                    throw new \LogicException('Extension "nesbot/carbon" has not been installed');
                }

                return Carbon::createFromFormat('Y-m-d H:i:s', (string) $value);
            case self::TYPE_TIMESTAMP_CARBON:
                if ($value === null || $value instanceof \DateTime) {
                    return $value;
                }

                if (!class_exists(Carbon::class)) {
                    throw new \LogicException('Extension "nesbot/carbon" has not been installed');
                }

                return (new Carbon())->setTimestamp((int) $value);
            default:
                throw new InvalidArgumentException("Unsupported attribute type '{$type}'");
        }
    }

    /**
     * Composes default value for {@see $attributeTypes} from the owner validation rules.
     *
     * @return array attribute type map.
     */
    protected function detectAttributeTypesFromRules(): array
    {
        $attributeTypes = [];
        foreach ($this->owner->getValidators() as $validator) {
            $type = null;
            if ($validator instanceof CBooleanValidator) {
                $type = self::TYPE_BOOLEAN;
            } elseif ($validator instanceof CNumberValidator) {
                $type = $validator->integerOnly ? self::TYPE_INTEGER : self::TYPE_FLOAT;
            } elseif ($validator instanceof CStringValidator) {
                $type = self::TYPE_STRING;
            }

            if ($type !== null) {
                $attributeTypes += array_fill_keys($validator->attributes, $type);
            }
        }

        return $attributeTypes;
    }

    /**
     * Detects attribute types from the owner's DB table schema.
     *
     * @return array<string, string> detected attribute types.
     */
    protected function detectAttributeTypesFromSchema(): array
    {
        $tableName = $this->owner->tableName();

        if (($table = $this->owner->getDbConnection()->getSchema()->getTable($tableName)) === null) {
            throw new CDbException(
                Yii::t('yii', 'The table "{table}" for active record class "{class}" cannot be found in the database.', [
                    '{class}' => get_class($this->owner),
                    '{table}' => $tableName,
                ])
            );
        }

        $attributeTypes = [];
        foreach($table->columns as $column) {
            $attributeTypes[$column->name] = $this->detectTypeFromDbColumnSchema($column);
        }

        return $attributeTypes;
    }

    /**
     * Detects the attribute type from DB column schema.
     *
     * @param \CDbColumnSchema $column DB column schema.
     * @return string type name.
     */
    protected function detectTypeFromDbColumnSchema(CDbColumnSchema $column): string
    {
        switch ($column->type) {
            case 'integer':
                return self::TYPE_INTEGER;
            case 'boolean':
                return self::TYPE_BOOLEAN;
            case 'double':
                return self::TYPE_FLOAT;
        }

        if (stripos($column->dbType, 'json') !== false) {
            return self::TYPE_ARRAY_OBJECT;
        }

        if (stripos($column->dbType, 'date') !== false) {
            return self::TYPE_DATETIME;
        }

        if (stripos($column->dbType, 'timestamp') !== false) {
            return self::TYPE_DATETIME;
        }

        return self::TYPE_STRING;
    }

    /**
     * Stashes original raw value of attribute for the future restoration.
     *
     * @param string $name attribute name.
     * @param mixed $value attribute raw value.
     * @return void
     */
    private function stashAttribute(string $name, $value): void
    {
        $this->_stashedAttributes[$name] = $value;
    }

    /**
     * Applies all stashed attribute values to the owner.
     *
     * @return void
     */
    private function applyStashedAttributes(): void
    {
        foreach ($this->_stashedAttributes as $name => $value) {
            $this->owner->setAttribute($name, $value);
            unset($this->_stashedAttributes[$name]);
        }
    }

    /**
     * Clears all stashed attributes.
     *
     * @return void
     */
    private function clearStashedAttributes(): void
    {
        $this->_stashedAttributes = [];
    }

    /**
     * Performs typecast for attributes values in the way they are suitable for the saving in database.
     * E.g. convert objects and arrays to scalars.
     *
     * @return void
     */
    protected function typecastAttributesForSaving(): void
    {
        $this->clearStashedAttributes();

        foreach ($this->owner->getAttributes() as $name => $value) {
            if ($value === null || is_scalar($value)) {
                continue;
            }

            if ($value instanceof \CDbExpression) {
                continue;
            }

            $this->stashAttribute($name, $value);

            if (is_array($value) || $value instanceof \JsonSerializable) {
                $this->owner->setAttribute($name, json_encode($value));

                continue;
            }

            if ($value instanceof \DateTime) {
                if (isset($this->attributeTypes[$name]) && $this->attributeTypes[$name] === self::TYPE_TIMESTAMP) {
                    $this->owner->setAttribute($name, $value->getTimestamp());
                } else {
                    $this->owner->setAttribute($name, $value->format('Y-m-d H:i:s'));
                }

                continue;
            }

            if ($value instanceof \Traversable) {
                $this->owner->setAttribute($name, json_encode(iterator_to_array($value)));

                continue;
            }

            $this->owner->setAttribute($name, (string) $value);
        }
    }

    // Event Handlers:

    /**
     * {@inheritdoc}
     */
    public function events(): array
    {
        $events = [];

        if ($this->typecastAfterValidate) {
            $events['onAfterValidate'] = 'afterValidate';
        }

        if ($this->getOwner() instanceof CActiveRecord) {
            $events['onBeforeSave'] = 'beforeSave';
            $events['onAfterSave'] = 'afterSave';

            if ($this->typecastAfterFind) {
                $events['onAfterFind'] = 'afterFind';
            }
        }

        return $events;
    }

    /**
     * Handles owner 'afterValidate' event, ensuring attribute typecasting.
     *
     * @param \CEvent $event event instance.
     */
    public function afterValidate(CEvent $event): void
    {
        if (!$this->owner->hasErrors()) {
            $this->typecastAttributes();
        }
    }

    /**
     * Handles owner 'beforeSave' owner event, ensuring attribute typecasting.
     *
     * @param \CModelEvent $event event instance.
     */
    public function beforeSave(CModelEvent $event): void
    {
        if ($this->typecastBeforeSave) {
            $this->typecastAttributes();
        }

        $this->typecastAttributesForSaving();
    }

    /**
     * Handles owner 'afterSave' event, ensuring attribute typecasting.
     *
     * @param \CEvent $event event instance.
     */
    public function afterSave(CEvent $event): void
    {
        $this->applyStashedAttributes();

        if ($this->typecastAfterSave) {
            $this->typecastAttributes();
        }
    }

    /**
     * Handles owner 'afterFind' event, ensuring attribute typecasting.
     *
     * @param \CEvent $event event instance.
     */
    public function afterFind(CEvent $event): void
    {
        $this->typecastAttributes();
    }
}